syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+
+	preinitScript = userScript = binName = imgName = NULL;
+
+	if (argc <= 1) {
+		sysgen_help(argv[0]);
+		return EXIT_FAILURE;
+	}
+
+	while ((opt = getopt(argc, argv, "s:p:u:b:i:h")) != -1) {
+		switch (opt) {
+			case 's':
+				if ((optind + 1) < argc) {
+					sysgen_common.pkernel = strtoul(argv[optind - 1], &endptr, 0);
+					if ((sysgen_common.pkernel == 0 && strlen(endptr) != 0) || (sysgen_common.pkernel != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong physical image address %s\n", argv[optind - 1]);
+						return -1;
+					}
+
+					sysgen_common.offs = strtoul(argv[optind], &endptr, 0);
+					if ((sysgen_common.offs == 0 && strlen(endptr) != 0) || (sysgen_common.offs != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage offset %s\n", argv[optind]);
+						return -1;
+					}
+
+					sysgen_common.maxsz = strtoul(argv[optind + 1], &endptr, 0);
+					if ((sysgen_common.maxsz == 0 && strlen(endptr) != 0) || (sysgen_common.maxsz != 0 && strlen(endptr) != 0)) {
+						fprintf(stderr, "Wrong syspage size %s\n", argv[optind + 1]);
+						return -1;
+					}
+				}
+				break;
+
+			case 'p':
+				preinitScript = optarg;
+				break;
+
+			case 'u':
+				userScript = optarg;
+				break;
+
+			case 'b':
+				binName = optarg;
+				break;
+
+			case 'i':
+				imgName = optarg;
+				break;
+
+			case 'h':
+			default:
+				sysgen_help(argv[0]);
+				return EXIT_SUCCESS;
+		}
+	}
+
+	sysgen_common.buff = calloc(sysgen_common.maxsz, sizeof(uint8_t));
+	if (sysgen_common.buff == NULL)
+		return EXIT_FAILURE;
+
+	sysgen_common.syspage = (syspage_t *)sysgen_common.buff;
+	sysgen_common.syspage->size = ALIGN_ADDR(sizeof(syspage_t), sizeof(long long));
+	sysgen_common.syspage->pkernel = sysgen_common.pkernel;
+	LIST_INIT(&sysgen_common.aliases);
+
+	/* Parse preinit script with map and console commands */
+	err = sysgen_parseScript(preinitScript);
+	if (err < 0) {
+		fprintf(stderr, "Cannot parse preinit script: %s\n", preinitScript);
+		sysgen_cleanup();
+		return err;
+	}
+
+	/* Parse user script descring programs in the syspage */
+	err = sysgen_parseScript(userScript);
+	if (err < 0) {
+		fprintf(stderr, "Cannot parse user script: %s\n", userScript);
+		sysgen_cleanup();
+		return err;
+	}
+
+	/* Based on scripts crate binary representation of syspage */
+	if (binName != NULL) {
+		err = sysgen_syspage2bin(binName);
+		if (err < 0) {
+			fprintf(stderr, "Cannot create binary syspage file: %s\n", binName);
+			sysgen_cleanup();
+			return err;
+		}
+		printf("Created binary syspage file: %s\n", binName);
+	}
+
+	/* Based on scripts write binary syspage directly to the image */
+	if (imgName != NULL) {
+		err = sysgen_addSyspage2Img(imgName);
+		if (err < 0) {
+			fprintf(stderr, "Cannot write binary syspage to kernel image: %s\n", imgName);
+			sysgen_cleanup();
+			return err;
+		}
+		printf("Syspage is written to image: %s at offset 0x%lx\n", imgName, sysgen_common.offs);
+	}
```

Comment:
Why 2 options?

Reply:
Initially I thought that such binary syspage blob can be used somewhere but probably it unnecessary. I have removed this option.